### PR TITLE
SPX 0 DTE 

### PR DESCRIPTION
**Pull request by Ininterrompue**
Implemented some nominal ranges. Uses the VIX directly, which assumes returns are normally distributed. This is NOT true of course, but this should be a reasonable start for an ML procedure. ML should also capture some of the empirically observed skew. For example, as of today, if you look at the options for the 1 sigma range for 5/2 expiring, the put side is 19 delta while the call side is 16 delta.

Also did a bit of preliminary feature engineering to play around with the data. Output:

```
            Date         Open         High          Low        Close    Adj Close      Volume      Change  % Change  abs(%ch)       Swing   % Swing        VIX
8139  2022-04-21  4489.169922  4512.939941  4384.470215  4393.660156  4393.660156  3833570000  -65.790039 -0.014753  0.014753  128.469726  0.028808  22.680000
8140  2022-04-22  4385.830078  4385.830078  4267.620117  4271.779785  4271.779785  3930660000 -121.880371 -0.027740  0.027740  118.209961  0.026905  28.209999
8141  2022-04-25  4255.339844  4299.020020  4200.819824  4296.120117  4296.120117  4061070000   24.340332  0.005698  0.005698   98.200196  0.022988  27.020000
8142  2022-04-26  4278.140137  4278.140137  4175.040039  4175.200195  4175.200195  3760420000 -120.919922 -0.028146  0.028146  103.100098  0.023998  33.520000
8143  2022-04-27  4186.520020  4240.709961  4162.899902  4183.959961  4183.959961  3876340000    8.759766  0.002098  0.002098   77.810059  0.018636  31.600000
------------------------------
SPX 4131.93, VIX 33.40
4044.99-4218.87 [68.27%]
3958.06-4305.80 [95.45%]
3871.12-4392.74 [99.73%]
```

### CODE REVIEW

**Comment by Ininterrompue** [comment]
Optimizer implemented, new output looks like this:
```
------------------------------
Nominal ranges:
SPX 4131.93, VIX 33.40
4044.99-4218.87 [84.13% one side, 68.27% both sides]
3958.06-4305.80 [97.72% one side, 95.45% both sides]
3871.12-4392.74 [99.87% one side, 99.73% both sides]
3988.93-4274.93 [95.00% one side, 90.00% both sides]
------------------------------
Running optimization algorithm...
...Complete.
------------------------------
POP threshold on one side: 95.00%
etaPut = 0.7388
etaCall = 0.7306
Multiplier = 1.6449
Actual success rates:
94.99% put side
95.00% call side
89.99% both sides
------------------------------
Empirical ranges:
SPX 4131.93, VIX 33.40
4026.28-4236.41 [95.00% one side, 90.00% both sides]
```

I'm a bit suspicious that the eta's are so far from 1...as in, if we tried using the ranges the optimizer is spitting out in the real world, we might get BTFO.

File: OptionPrice.py [hunk 1]
```diff
@@ -91,35 +91,35 @@ def getDiscount(self):
 
     def getOptionPrice(self):
         callPrice = self.N(self.d1) * self.S - self.N(self.d2) * self.K * self.discount
-        if self.optionType.value == 'call':
+        if self.optionType is OptionType.CALL:
             return callPrice
-        elif self.optionType.value == 'put':
+        elif self.optionType is OptionType.PUT:
```
Comment (mikinty) [inline OptionPrice.py:96]: Normally we don't like to make "unrelated changes" in our PRs, meaning if you PR is to implement some algorithm, we don't want to see you do other stuff in that PR as well, like in this case fix the enum comparison styles. But honestly this is fine for now, just something to keep in mind for the future.

File: SPXIronCondors.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+SPX 4 DTE backtesting: Monday -> Friday
+'''
+import pandas as pd
+import numpy as np
+from scipy.stats import norm
+from scipy.optimize import minimize
+from util.Enums import Ticker, Columns
+from OptionPrice import Constants
+
+class SPX0DTE:
```
Comment (mikinty) [inline SPXIronCondors.py:13]: This is fine for now, but I imagine we'd like to implement a "N" DTE strategy, so having an N DTE class in the constructor might be useful

File: SPXIronCondors.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    '''
+    Let S[n] be today's closing price on SPX and S[n+1] be tomorrow's closing price.
+
+    Let p be the probability of profit (pop) threshold on one side of the iron condor.
```
Comment (mikinty) [inline SPXIronCondors.py:17]:
```suggestion
    Let p be the probability of profit (PoP) threshold on one side of the iron condor.
```

Nit: I think you were using PoP later, but this is whatever.

Comment (mikinty) [reply SPXIronCondors.py:17]: I probably prefer P(profit), but this is fine

File: SPXIronCondors.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+        self.dfSPX = self.parseTicker(Ticker.SPX)
+        self.dfVIX = self.parseTicker(Ticker.VIX)
+
+        # define outputs
+        self.etaPut  = 1
+        self.etaCall = 1
+        
+    def checkPopThreshold(self):
+        if self.popThreshold >= 1.0 or self.popThreshold <= 0.0:
```
Comment (mikinty) [inline SPXIronCondors.py:73]: Isn't == 0 or 1.0 fine? Shouldn't it be non-equal bounds here?

**Review by mikinty — APPROVED** [review]
lgtm, very nice :)

File: OptionPrice.py (hunk 1 above)
Comment (Ininterrompue) [reply OptionPrice.py:96]: Got it, yeah this is a bad habit of mine

File: SPXIronCondors.py (hunk 3 above)
Comment (Ininterrompue) [reply SPXIronCondors.py:17]: PoP is the standard acronym

File: SPXIronCondors.py (hunk 4 above)
Comment (Ininterrompue) [reply SPXIronCondors.py:73]: No, ppf(0) or ppf(1) is evaluated in __init__ and those are infinite. Switched the check function with the popMultiplier to make the logic more straightforward.

File: SPXIronCondors.py (hunk 2 above)
Comment (Ininterrompue) [reply SPXIronCondors.py:13]: Will be on the to-do list. Only caution is that, for example 4 DTE, we have to pay attention to only holding positions Monday to Friday - no holding over weekends.

**Pull request merged**